### DISCUSSION TITLE
Can you really compute a short-term average rating across agencies?

### DISCUSSION THREAD

**Discussion by a-vester** [hsbc/pyratings#18] [General]
``pyratings`` current approach to short-term ratings is to make them comparable across rating agencies (same philosophy as for long-term ratings). This is a pre-requisite in order to compute an average rating on a portfolio basis, where different securities have ratings from different agencies.  
However, compared to long-term ratings, there is a significant difference. Different rating agencies use a different number of ratings. For example, while Moody`s uses four short-term ratings, S&P and Fitch use seven, and DBRS uses ten.

| Moody’s | Score |  S&P | Score | Fitch | Score |    DBRS    | Score |
|:-------:|------:|:----:|------:|:-----:|------:|:----------:|------:|
|   P-1   |     1 | A-1+ |     1 |  F1+  |     1 | R-1 (High) |     1 |
|   P-2   |     2 |  A-1 |     2 |   F1  |     2 |  R-1 (Mid) |     2 |
|   P-3   |     3 |  A-2 |     3 |   F2  |     3 |  R-1 (Low) |     3 |
|    NP   |     4 |  A-3 |     4 |   F3  |     4 | R-2 (High) |     4 |
|         |       |   B  |     5 |   B   |     5 |  R-2 (Mid) |     5 |
|         |       |   C  |     6 |   C   |     6 |  R-2 (Low) |     6 |
|         |       |   D  |     7 |   D   |     7 |     R-3    |     7 |
|         |       |  SD  |     7 |   RD  |     7 |     R-4    |     8 |
|         |       |      |       |       |       |     R-5    |     9 |
|         |       |      |       |       |       |      D     |    10 |

![image](https://user-images.githubusercontent.com/104636025/199541410-111a8c46-83ed-4596-9ef6-d1384556ea91.png)
Source: https://www.moneyland.ch/en/rating-agencies

As these rating agencies don't necessarily share the same scaling, it seems to be somewhat questionable if it is reasonable to compute an average short-term credit rating across multiple agencies at all?

**Comment by sander2825** [comment]
Hi Andreas, 

 What if you apply the LT ratings scaling for ST ratings but introduce an "average equivalent value" or a "{min equivalent value, max equivalent value} list"?

For instance, let's consider a company rated P-1, A-1+and F1/F1+ by respectively Moody's, S&P and Fitch.

If you consider an average equivalent value, the ratings would lead to respectively 3 (as P-1 is "equivalent" to a LT rating between AAA (ie 1) and A1 (ie 5)), 2.5 (as A-1+ is  "equivalent" to a LT rating between AAA (ie 1) and AA- (ie 4)) and 5 (as F1/F1+ is "equivalent" to A+ LT rating). You would then be able to compute a ST average rating for this company of 3.5 (equivalent to a AA/AA- in LT rating scale).

If you consider a {min, max} list, the ratings of the company would be translated into respectively {1,5}, {1,4} and {5,5} and you may compute a ST average rating for this company of {2.3, 4.7}. 
Even if more complicated, having a list would help to keep track of the amplitude of the ST rating scaling with respect to the LT rating.

**Comment by a-vester** [comment]
Hi Sander,

Thank you very much for your suggestion!

Let's recap the process of computing an average **long-term** rating for a portfolio consisting of multiple securities. Each security **may** have a rating from different rating agencies.

Perform the following steps:
1. Clean ratings, i.e. remove watches/outlooks.
2. Consolidate ratings, i.e. compute, let's say, the **worst rating** for every security by choosing a rating agency's scale for "displaying" the rating.
3. Translate consolidated rating into rating score.
4. Compute the weighted average score (sum-product of security weights and scores).
5. Translate average score back to human readable rating by choosing a rating agency's scale for "displaying" the rating.

With respect to **short-term** ratings, we might experience some difficulties.
Step 2:  
Consolidating the ratings is more difficult as the rating scales don't match among different agencies. If we were to compute the worst rating for every security, we need to choose an agency's ratings scale to display it in the worst_rating column. How would you do that?  
If we use your "average equivalent value" and determine the maximum value for each security, we directly get a rating score. But what if I wanted to express the worst rating in terms of S&P's rating scale? Going with your example, the value 5 would then translate to A-1, wouldn't it?  
Using your [min, max] list would result in 4.7, which, if rounded to 5, lead to the same conclusion.

**Comment by MarcoGER** [comment]
Hi all,

I think I agree with Sander (if I understand him correctly): for each ST rating you can find an equivalent LT rating. The LT ratings among different rating agencies match again. So if you have two rating agencies, you could go from ST(A) -> LT(A) -> LT(B) -> ST(B)
So for Moddys e.g. I would match the following numbers:
P-1 : 3 (Middle of Aaa to A1)
P-2 : 7 (A2 to Baa1)
etc.
Question is how exactly you match e.g. P-1: whether you consider the range from Aaa to A1 or Aaa to Baa1.

In general it may make sense to use a parameter in case you find different methods for the calculations so that the user can select a different option.

Does this make sense?

**Comment by sander2825** [comment]
Hi all, 
It’s indeed a good idea to use a parameter for specifying the method (as there is probably no consensus on how to mix and compare ST ratings). Furthermore, you are right Marco, I have forgotten about the overlap. Probably we shall consider P-1 between Aaa and A3 (i.e. a grade between 1 and 7, and thus an average grade of 4), P-2 between A2 and Baa2 (a grade between 6 and 9 -> 7.5), and so on…

Let me clarify with an example, according to the details that Andreas gave: the idea would need to invert steps 2 and 3; and to split the 5th step into “translate to LT rating” and then “translate to ST rating”.

Let’s consider a portfolio with two securities (90% in sec. A/10% in sec. B). Here would be the pipeline :

![image](https://user-images.githubusercontent.com/67168509/200354273-e4564e02-2851-479b-a09d-12f7e1dba19b.png)




I am nevertheless wondering about another option: if the moneyland table is correct and if most (if not all) ST issuers have also a LT rating, why don’t we use LT ratings only and transpose the final output in ST equivalent according to the moneyland table ? This would limit the error embedded in ST rating (a ST P-1  bond can either be Aaa or A3 in LT). Here would then be the pipeline :

![image](https://user-images.githubusercontent.com/67168509/200354351-b8f35c81-9476-4f9a-8cbc-9438d8c8ac09.png)

**Reply by a-vester** [reply]
@sander2825 
@MarcoGER 
Thank you for your input. This is really helpful.

Before replying to your suggestions, I think it makes sense to take a closer look at how short-term ratings work.

Short-term ratings are highly dependent on long-term ratings. However, they take into account the short-term prospects of an issuer from a liquidity perspective, i.e. the short-term rating may be better or worse than what could be implied by simply translating the long-term rating. The table below shows the possible variations.
 
For example, an issuer rated **A3** would usually get assigned a short-term rating of **P-2**. However, if the liquidity prospects are better than what a **P-2** rating would imply (e.g. the issuer might have some guaranties from other entities), a **P-1** rating might be assigned.  
On the other hand, an issuer rated **A2** would normally get assigned a short-term rating of **P-1**. However, if the rating agency concludes that the liquidity prospects are worse than what a **P-1** rating usually implies, the issuer might get assigned a rating of **P-2**.

As a result, an issuer can only have one rating at a time. There is no such thing as a split rating (from the same rating agency).

![image](https://user-images.githubusercontent.com/104636025/201066265-9f9eb969-67b1-4c1c-a24c-5017557be481.png)

With that in mind, I think @sander2825's second suggestion might be a bit problematic. As pointed out above, an issuer with an **A2** long-term rating should usually have a short-term rating of **P-1**. However, this is not guaranteed. As a result, simply using the long-term ratings might lead to skewed results.

I am therefore more inclined to follow on with @sander2825 's first suggestion, namely assigning an average equivalent value (``AEV``). When computing the average, I tend to use the "regular" ranges as described as per the below table. In my view, the advantage would be to have clear cuts between ratings. The only exception is DBRS' **R-2 L** / **R-3** rating. 

![image](https://user-images.githubusercontent.com/104636025/201072058-92f55511-47b6-4b40-98a8-640a6c7c3e46.png)

The two security example should then look as follows:

![image](https://user-images.githubusercontent.com/104636025/201071867-6ce26983-652e-4d18-b380-a2cd815c68fc.png)

**Comment by MarcoGER** [comment]
Hey guys,

I would agree to what you have both said: Sander's first suggestion is what I would also think makes most sense to me.
The motivation I think is the following: you cannot directly (!) compare two ST ratings from two different rating agencies. But what exists is:
1. A relationship between LT ratings for two different rating agencies (such as Aaa is the same as AAA which all means "1").
2. A relationship between LT and ST rating WITHIN one rating agency (such as A-1+ is the "middle of AAA and AA-").

So what you actually do is go from ST1 to LT1 in rating agency A, then move over to rating agency B (via relationship 1 above), and within rating agency B from from LT2 to ST2 (via relationship 2).

And that is exactly what you both are doing in your charts: by assigning a number to a ST rating within one rating agency (e.g. the 2.5 for A-1+), you indirectly move from the ST rating to the LT equivalent rating (I mean why is A-1+ equivalent to 2.5 and not 10?) If you have that, you can compare those numbers again for all LT ratings among rating agencies, before you then move back to the equivalent ST rating.

Hope my thoughts make any sense, but in short: Sander's option 1 is my favorite ;-)

**Comment by sander2825** [comment]
Hi! 

Thanks for the feedback!

@a-vester your explanation on ST ratings makes sense and indeed my second proposal is not in line with the way ST ratings are created but the idea was to limit the "dispersion" between ratings; i.e. an A2 issuer would be 6 and this value means that it is either a "low" P-1 or a "high "P-2". But it's true that we loose the real ST rating information. So let's not consider this idea.

I understand your idea of getting rid of the overlap, but it's a choice as well to do so. 
What if these two mappings are implemented (the first I proposed and the one with no overlapping) with a parameter on the function (as @MarcoGER proposed) to decide which one is chosen?

**Reply by a-vester** [reply]
> I understand your idea of getting rid of the overlap, but it's a choice as well to do so.

@sander2825 and @MarcoGER You're right. The user should be able to choose the methodology when translating short term ratings into scores.

Here's a suggestion that takes into account three different methodologies:

1. **Best**: Uses the best possible short-term ratings when translating to a long-term rating and vice versa.
2. **Avg**: Uses the most likely short-term ratings when translating to a long-term rating and vice versa.
3. **Worst**: Uses the worst possible short-term ratings when translating to a long-term rating and vice versa.

To explain it, here's a snippet from my table (see above):

![image](https://user-images.githubusercontent.com/104636025/202234180-326678a1-5c9b-41c7-8242-ab6716c170d5.png)

1. **Best**:
    * P-1 has a range from Aaa to A3 (score 1 to 7)
    * P-2 has a range from Baa1 to Baa2 (score 8 to 9)
    * P-3 has a range from Baa3 to Baa3 (score 10)
    * NP has a range from Ba1 to D (score 11 to 22)
2. **Avg**:
    * P-1 has a range from Aaa to A2 (score 1 to 6)
    * P-2 has a range from A3 to Baa1 (score 7 to 8)
    * P-3 has a range from Baa2 to Baa3 (score 9 to 10)
    * NP has a range from Ba1 to D (score 11 to 22)
3. **Worst**:
    * P-1 has a range from Aaa to A1 (score 1 to 5)
    * P-2 has a range from A2 to Baa1 (score 6 to 8)
    * P-3 has a range from Baa2 to Baa3 (score 9 to 10)
    * NP has a range from Ba1 to D (score 11 to 22)

Here is how it looks like for all rating agencies:

![image](https://user-images.githubusercontent.com/104636025/202218274-0d27c1c5-df02-4771-916d-4d0daa4af3a6.png)

![image](https://user-images.githubusercontent.com/104636025/202231326-e18dbecf-616f-434b-9648-96d37c4e7e98.png)

![image](https://user-images.githubusercontent.com/104636025/202231425-b3769ec1-6840-4cec-862c-a199e34732be.png)

![image](https://user-images.githubusercontent.com/104636025/202231508-7cdd0b3b-61be-4b9f-85e6-eb5990288825.png)

![image](https://user-images.githubusercontent.com/104636025/202235755-42c498f7-7459-45c7-bea9-831e0cd10c4d.png)

One of the advantages of using different methodologies is to not have to handle any overlapping ranges. Every short-term rating has a one-to-one relationship with an equivalent long-term rating.

**Comment by sander2825** [comment]
Great!
